### PR TITLE
issue #165 - better discoverability for packs

### PR DESCRIPTION
**Pull request by pounard**
See issue #165

### CODE REVIEW

File: src/Anonymization/Anonymizer/AnonymizerRegistry.php [hunk 1]
```diff
@@ -4,134 +4,233 @@
...
+        IbanBicAnonymizer::class,
+        IntegerAnonymizer::class,
+        LastNameAnonymizer::class,
+        LoremIpsumAnonymizer::class,
+        Md5Anonymizer::class,
+        NullAnonymizer::class,
+        PasswordAnonymizer::class,
+        StringAnonymizer::class,
+    ];
+
```
Comment (SimonMellerin) [inline src/Anonymization/Anonymizer/AnonymizerRegistry.php:42]:
```suggestion
use MakinaCorpus\DbToolsBundle\Anonymization\Anonymizer\Core;

class AnonymizerRegistry
{
    // Core anonymizers list is hardcoded.
    private static $coreAnonymizers = [
        Core\AddressAnonymizer::class,
        Core\ConstantAnonymizer::class,
        Core\DateAnonymizer::class,
        Core\EmailAnonymizer::class,
        Core\FirstNameAnonymizer::class,
        Core\FloatAnonymizer::class,
        Core\IbanBicAnonymizer::class,
        Core\IntegerAnonymizer::class,
        Core\LastNameAnonymizer::class,
        Core\LoremIpsumAnonymizer::class,
        Core\Md5Anonymizer::class,
        Core\NullAnonymizer::class,
        Core\PasswordAnonymizer::class,
        Core\StringAnonymizer::class,
    ];

```

File: src/Anonymization/Anonymizer/AnonymizerRegistry.php [hunk 2]
```diff
@@ -4,134 +4,233 @@
...
+    public function __construct(?array $paths = null)
+    {
+        $this->addPath($paths ?? []);
     }
 
-    public function addPath(array $paths): void
+    /**
+     * Add path in which to lookup for anonymizers.
+     */
+    public function addPath(array|string $paths): void
```
Comment (SimonMellerin) [inline src/Anonymization/Anonymizer/AnonymizerRegistry.php:54]: I don't get why this has to be changed ? Is this not the exact same thing ?

I think we can keep the original signature

File: src/Anonymization/Anonymizer/AnonymizerRegistry.php [hunk 3]
```diff
@@ -4,134 +4,233 @@
...
+
+            $ignoreList = '@(composer|doctrine|nikic|monolog|phar-io|phpunit|phpstan|psr|sebastian|symfony|twig)/*@';
 
-        return $this->getAnonymizers()[$name];
+            foreach (InstalledVersions::getInstalledPackages() as $package) {
+                if (\preg_match($ignoreList, $package)) {
+                    continue;
+                }
+
+                $directory = InstalledVersions::getInstallPath($package);
```
Comment (SimonMellerin) [inline src/Anonymization/Anonymizer/AnonymizerRegistry.php:190]: with `InstalledVersions::getInstalledPackages` you could avoid to parse `composer.json` files below

File: src/Anonymization/Anonymizer/AnonymizerRegistry.php (hunk 1 above)
Comment (pounard) [reply src/Anonymization/Anonymizer/AnonymizerRegistry.php:42]: OK why not.

File: src/Anonymization/Anonymizer/AnonymizerRegistry.php (hunk 2 above)
Comment (pounard) [reply src/Anonymization/Anonymizer/AnonymizerRegistry.php:54]: I kept `array` for backward compatibility, but in `locatePacks` I simply add strings.

File: src/Anonymization/Anonymizer/AnonymizerRegistry.php (hunk 3 above)
Comment (pounard) [reply src/Anonymization/Anonymizer/AnonymizerRegistry.php:183]: Done !

File: src/Anonymization/Anonymizer/AnonymizerRegistry.php (hunk 1 above)
Comment (pounard) [reply src/Anonymization/Anonymizer/AnonymizerRegistry.php]: Done !

**Comment by pounard** [comment]
Just added a changelog line as well.

**Comment by pounard** [comment]
Again and again and again the same random failure with the same version of MySQL...

**Review by SimonMellerin — APPROVED** [review]

**Pull request merged**